GameServerLib/GameObjects/Other/CollisionHandler.cs

```diff
@@ -44,7 +81,7 @@ public void Update()
                 {
                     obj.OnCollision(null);
                 }
-
+                // TODO: better algos
```

Comment:
I reckon a KD tree, or simply subdividing the map into multiple sub-sections would help tremendously.

Reply:
Probably. If you want you can try to implement it. But I think this is an overkill.

Reply:
It's actually probably the best way to optimise this. You don't want to check collisions with objects on the other side of the map.

Reply:
Where would that occur with this implementation?

Reply:
Unless I'm mistaken, _objects is a list of all the game objects in the game. It goes through two forloops, checking each game object against eachother, for the whole map.

Reply:
So we need to consider using lib like https://www.nuget.org/packages/KdTree/, and generating some scheme for static objects like:
Generally if we have 2 circles in 2D the distance between them is $\sqrt{(x_1-x_2)^2+(y_1-y_2)^2}-r_1-r_2$. Basically, if we search for collision we only care about to minimize the $\sqrt{(x_1-x_2)^2+(y_1-y_2)^2}-r_{static}$ part. The problem is that is no longer positive and so it isn't a metric. One option is heuristic search based on nearest points and radius. Elevating to 3D, we can do something that minimizes $\sqrt{(x_1-x_2)^2+(y_1-y_2)^2+r_{static}^2}<\sqrt{(x_1-x_2)^2+(y_1-y_2)^2}+r_{static}$ but greater than $\sqrt{(x_1-x_2)^2+(y_1-y_2)^2}-r_{static}$. I don't think that solves this problem. If we can do that in some way we have O(log(n)) algo, but I think it will be easier to have our O(n) check between the dynamic and all the static ones. (of course we can try kd tree with stop criterion when the NN is at least max radius + dynamic radius away, if we think most radius are the same)

For general algo, we can think about checking every dynamic object against the others using some region separation. (max radius + dynamic radius away and iterating the max radius from this range again and so)